File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/importering/autoimport/statistikk/utils/SykefraværsstatistikkLandUtils.java
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.importering.autoimport.statistikk.utils;

import no.nav.arbeidsgiver.sykefravarsstatistikk.api.importering.Sykefraværsstatistikk;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.ÅrstallOgKvartal;
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
import org.springframework.jdbc.core.namedparam.SqlParameterSource;
import org.springframework.jdbc.core.namedparam.SqlParameterSourceUtils;

import java.util.Arrays;
import java.util.List;

public class SykefraværsstatistikkLandUtils extends SykefraværsstatistikkIntegrasjon
        implements SykefraværsstatistikkIntegrasjonUtils {


    public SykefraværsstatistikkLandUtils(NamedParameterJdbcTemplate namedParameterJdbcTemplate) {
        super(namedParameterJdbcTemplate);
    }

    @Override
    public DeleteSykefraværsstatistikkFunction getDeleteFunction() {
        DeleteSykefraværsstatistikkFunction function =
                (ÅrstallOgKvartal årstallOgKvartal) -> {
                    SqlParameterSource namedParameters =
                            new MapSqlParameterSource()
                                    .addValue(ARSTALL, årstallOgKvartal.getÅrstall())
                                    .addValue(KVARTAL, årstallOgKvartal.getKvartal());

                    int antallSlettet =
                            namedParameterJdbcTemplate.update(
                                    String.format(
                                            "delete from sykefravar_statistikk_land where arstall = :%s and kvartal = :%s",
                                            ARSTALL, KVARTAL),
                                    namedParameters);
                    return antallSlettet;
                };
        return function;
    }

    @Override
    public BatchCreateSykefraværsstatistikkFunction getBatchCreateFunction(
            List<? extends Sykefraværsstatistikk> statistikk
    ) {

        BatchCreateSykefraværsstatistikkFunction function =
                () -> {
                    SqlParameterSource[] batch = SqlParameterSourceUtils.createBatch(statistikk.toArray());

                    int[] results = namedParameterJdbcTemplate.batchUpdate(
                            "insert into sykefravar_statistikk_land " +
                                    "(arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) " +
                                    "values " +
                                    "(:årstall, :kvartal, :antallPersoner, :tapteDagsverk, :muligeDagsverk)",
                            batch);
                    return Arrays.stream(results).sum();
                };

        return function;
    }

}

File: src/main/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/infrastruktur/database/statistikk/SykefraværsstatistikkLandUtils.kt
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.infrastruktur.database.statistikk

import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.Sykefraværsstatistikk
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.applikasjon.domenemodeller.ÅrstallOgKvartal
import org.springframework.jdbc.core.namedparam.MapSqlParameterSource
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.jdbc.core.namedparam.SqlParameterSource
import org.springframework.jdbc.core.namedparam.SqlParameterSourceUtils

class SykefraværsstatistikkLandUtils(namedParameterJdbcTemplate: NamedParameterJdbcTemplate) :
    SykefraværsstatistikkIntegrasjon(namedParameterJdbcTemplate), SykefraværsstatistikkIntegrasjonUtils {
    override fun getDeleteFunction(): DeleteSykefraværsstatistikkFunction {
        return DeleteSykefraværsstatistikkFunction { (årstall, kvartal): ÅrstallOgKvartal ->
            val namedParameters: SqlParameterSource = MapSqlParameterSource()
                .addValue(ARSTALL, årstall)
                .addValue(KVARTAL, kvartal)
            val antallSlettet = namedParameterJdbcTemplate.update(
                // language=postgresql
                "delete from sykefravar_statistikk_land where arstall = :$ARSTALL and kvartal = :$KVARTAL",
                namedParameters
            )
            antallSlettet
        }
    }

    override fun getBatchCreateFunction(
        list: List<Sykefraværsstatistikk>
    ): BatchCreateSykefraværsstatistikkFunction {
        return BatchCreateSykefraværsstatistikkFunction {
            val batch =
                SqlParameterSourceUtils.createBatch(*list.toTypedArray<Sykefraværsstatistikk?>())
            namedParameterJdbcTemplate.batchUpdate(
                "insert into sykefravar_statistikk_land (arstall, kvartal, antall_personer, tapte_dagsverk, mulige_dagsverk) values (:årstall, :kvartal, :antallPersoner, :tapteDagsverk, :muligeDagsverk)",
                batch
            ).sum()
        }
    }
}
